### ISSUE TITLE
Context-aware predicates in serializers

### ISSUE THREAD

**Issue by christhekeele** [rails-api/active_model_serializers#801]
Re: PR #498, I'd like to open up a discussion on a new potential DSL for serializers.

I've gotten a lot of milage out of this library in the past back by hacking in context-awareness into serializer instances. I'd love it if they were a core feature.

At any moment in the serialization process, a serializer ought to be able to know where it's at in relation to the grand scheme of things. This opens up a few cool options: condensing multiple serializers into one, easily avoiding expensive sub-serializations, etc.

As an example, the DSL I've made for myself behaves as follows:

```ruby
class ContextAwareSerializer < ActiveModel::Serializer

  def initialize(object, options={})
    @associated   = options[:associated]
    @listed       = options[:listed]
    @show_details = options[:details]
    super
  end

  def associated?; @associated; end
  def listed?;     @listed; end

  def nested? listed? or associated?; end
  def show_details?
    if @show_details or @show_details.nil? # If not ~explicitly~ set to false,
      @show_details or not nested?         # only show details if asked for, or at top level
    end
  end

  def filter(keys)
    if show_details?
      super
    else
      super - detail_keys
    end
  end

end
  
class UserSerializer < ContextAwareSerializer

  attributes :id, :uri, :name
  detailed attributes :first_name, :last_name, :bio

  has_one :account
  has_many :posts
  detailed has_many :comments

end
```

I've made it such that serializers instantiated by `ActiveModel::ArraySerializer` receive `options[:listed]`,`ActiveModel::Serializer::Association` passes in `options[:associated]`, and the standard custom options from the controller provide the `options[:details]`.

I'd like to propose incorporating the `nested` DSL: adding the injection of and predicates for `listed?`, `associated?`, and `nested?`.

I think the `detailed` DSL requires more consideration from other people's perspectives and usecases. Certainly, it can be far more easily hacked in than the nested DSL, since you don't have to monkey patch any `active_model_serializer` classes to achieve it. Perhaps it merits a place in the library as a subclass, or disabled-by-default feature. YMMV.

**Comment by christhekeele** [comment]
My monkeypatch implementation of this feature has made its way through my `config/initializers/active_model_serializers.rb` in various projects for almost 2 years now, since `v0.8.0`, and since I've recently updated it to be `0.10.0` compatible, I thought I'd bump it.

I'm interested to hear thoughts on this interface and API, but since there's not been any discussion to date, I've furnished a master-mergeable `0.10.0.rc1` PR with my work to date as a starting point.

TLDR:

- `ArraySerializers` inject `listed: true` into `ActiveModel::Serializers` they initialize
- `ActiveModel::Serializers` inject `associated: true` into other `ActiveModel::Serializers` they initialize
- `ActiveModel::Serializers` register these options with instance variables which are consumed in the `listed?`, `associated?`, and conjunctive `nested?` predicate methods
- Library users can leverage them by switching on these predicates in attribute/association method overrides

This DSL introduces no breaking changes and is entirely opt-in, although the implications on the caching mechanism merits a closer look.
